source/SkiaSharp.HarfBuzz/SkiaSharp.HarfBuzz/CanvasExtensions.cs

```diff
@@ -99,7 +104,89 @@ public static void DrawShapedText(this SKCanvas canvas, SKShaper shaper, string
 			}
 
 			// draw the text
-			canvas.DrawText(textBlob, xOffset, 0, paint);
+			canvas.DrawText(textBlob, x + xOffset, y, paint);
+
+			if (clearCacheTimer is null && cacheDuration > 0)
+				clearCacheTimer = new Timer(_ => ClearCache(), null, 0, cacheDuration);
+		}
+
+		private static uint cacheDuration = 0;
+
+		public static void SetShaperCacheDuration(this SKCanvas canvas, uint milliseconds) => SetShaperCacheDuration(milliseconds);
+		public static void SetShaperCacheDuration(uint milliseconds)
+		{
+			cacheDuration = milliseconds;
+
+			clearCacheTimer?.Change(0, cacheDuration);
+		}
+
+		private static readonly Dictionary<int, (SKShaper shaper, DateTime cachedAt)> shaperCache = new();
+		private static readonly Dictionary<int, (SKShaper.Result shapeResult, DateTime cachedAt)> shapeResultCache = new();
+
+		private static SKShaper GetShaper(SKTypeface typeface)
+		{
+			if (cacheDuration == 0)
+				return new SKShaper(typeface);
+
+			var key = HashCode.Combine(typeface.FamilyName, typeface.IsBold, typeface.IsItalic);
+
+			SKShaper shaper;
+			lock (shaperCache)
+			{
+				shaper = shaperCache.TryGetValue(key, out var value)
+					? value.shaper
+					: new SKShaper(typeface);
+
+				shaperCache[key] = (shaper, DateTime.Now);      // update timestamp
```

Comment:
Can this not be done in a single step using `GetOrAdd`?

https://learn.microsoft.com/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=net-8.0#system-collections-concurrent-concurrentdictionary-2-getoradd(-0-system-func((-0-1)))

Reply:
I don't know how `GetOrAdd` could help, but I tried `AddOrUpdate` [here](https://github.com/MichaelRumpler/SkiaTextRendering/blob/f937f9c459c79cfdbc50a9290a010cdeb31936d8/SkiaTextRendering/Extensions/CanvasExtensionsConcurrentDict.cs#L98):

	var newTuple = shaperCache.AddOrUpdate(key,
		_ => (new SKShaper(typeface), DateTime.Now),
		(_, old) => (old.shaper, DateTime.Now));

But the [simple](https://github.com/MichaelRumpler/SkiaTextRendering/blob/f937f9c459c79cfdbc50a9290a010cdeb31936d8/SkiaTextRendering/Extensions/CanvasExtensions.cs#L99C1-L107C1) `Dictionary` with a `lock` is still faster.

	lock (shaperCache)
	{
		shaper = shaperCache.TryGetValue(key, out var value)
			? value.shaper
			: new SKShaper(typeface);

		shaperCache[key] = (shaper, DateTime.Now);      // update timestamp
	}